marconi-chain-index/src/Marconi/ChainIndex/Indexers.hs

```diff
@@ -450,3 +477,29 @@ runIndexers socketPath networkId cliChainPoint traceName list = do
               <+> pretty chainPoints <> "."
               <+> "Please check the slot number and the block hash do belong to the chain"
     in io `catch` handleException
+
+toException :: Exception err => ExceptT err IO a -> IO a
+toException mx = do
+    x <- runExceptT mx
+    case x of
+        Left err  -> throw err
+        Right res -> pure res
+
+updateWith
+    :: MVar a
+    -> MVar err
+    -> (a -> ExceptT err IO a)
+    -> IO a
+updateWith xBox errBox f = modifyMVar xBox $ \x -> do
+  res <- runExceptT $ f x
+  case res of
+      Left err -> do
+          tryPutMVar errBox err $> (x, x)
+      Right x' -> pure (x', x')
+
+failWhenFull :: MonadIO m => MVar a -> m ()
+failWhenFull x = do
+  isEmpty <- liftIO $ isEmptyMVar x
+  if isEmpty
+  then pure ()
+  else error "an indexer raised an error"
```

Comment:
I still don't like to exit workers with an error, but doing proper thread management is a bit of work: we need to gather the `ThreadId`, keep them in the `Coordinator` and use them to kill the threads on error. With the current design, it's another significant refactoring.